src/java.base/share/classes/jdk/internal/lang/stable/StableValueImpl.java

```diff
@@ -0,0 +1,216 @@
+    private StableValueImpl() {}
+
+    @ForceInline
+    @Override
+    public boolean trySet(T content) {
+        if (wrappedContentAcquire() != null) {
+            return false;
+        }
+        // Prevent reentry via an orElseSet(supplier)
+        preventReentry();
+        // Mutual exclusion is required here as `orElseSet` might also
+        // attempt to modify the `wrappedValue`
+        synchronized (this) {
+            return wrapAndSet(content);
+        }
+    }
+
+    @ForceInline
+    @Override
+    public void setOrThrow(T content) {
+        if (!trySet(content)) {
+            // Neither the set content nor the provided content is reveled in the
+            // exception message as it might be sensitive.
+            throw new IllegalStateException("The content is already set");
+        }
+    }
+
+    @ForceInline
+    @Override
+    public T orElseThrow() {
+        final Object t = wrappedContentAcquire();
+        if (t == null) {
+            throw new NoSuchElementException("No content set");
+        }
+        return unwrap(t);
+    }
+
+    @ForceInline
+    @Override
+    public T orElse(T other) {
+        final Object t = wrappedContentAcquire();
+        return (t == null) ? other : unwrap(t);
+    }
+
+    @ForceInline
+    @Override
+    public boolean isSet() {
+        return wrappedContentAcquire() != null;
+    }
+
+    @ForceInline
+    @Override
+    public T orElseSet(Supplier<? extends T> supplier) {
+        Objects.requireNonNull(supplier);
+        final Object t = wrappedContentAcquire();
+        return (t == null) ? orElseSetSlowPath(supplier) : unwrap(t);
+    }
+
+    @DontInline
+    private T orElseSetSlowPath(Supplier<? extends T> supplier) {
+        preventReentry();
+        synchronized (this) {
+            final Object t = content;  // Plain semantics suffice here
+            if (t == null) {
+                final T newValue = supplier.get();
+                // The mutex is not reentrant so we know newValue should be returned
+                wrapAndSet(newValue);
+                return newValue;
+            }
+            return unwrap(t);
+        }
+    }
+
+    // The methods equals() and hashCode() should be based on identity (defaults from Object)
+
+    @Override
+    public String toString() {
+        final Object t = wrappedContentAcquire();
+        return t == this
+                ? "(this StableValue)"
+                : renderWrapped(t);
+    }
+
+    // Internal methods shared with other internal classes
+
+    @ForceInline
+    public Object wrappedContentAcquire() {
+        return UNSAFE.getReferenceAcquire(this, CONTENT_OFFSET);
+    }
+
+    static String renderWrapped(Object t) {
+        return (t == null) ? UNSET_LABEL : Objects.toString(unwrap(t));
+    }
+
+    // Private methods
+
+    private void preventReentry() {
+        if (Thread.holdsLock(this)) {
+            throw new IllegalStateException("Recursive initialization is not supported");
```

Comment:
Prevent users from having to look through the stack trace. Also, 'supported' sounds like it might be implemented later. I think 'allowed' is better to signal that this is an invariant.

```suggestion
            throw new IllegalStateException("Recursive initialization of a StableValue is not allowed");
```

Reply:
Or maybe a message like "Illegal recursive initialization of stable value".